### PR TITLE
Roadmap update

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: content/pages/roadmap.html [hunk 1]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
 
-                <p>Each part of the file will be transferred and tracked as a regular data object within the iRODS protocol.  Any individual part that does not transfer correctly can be retransmitted.  This will allow iRODS to more gracefully overcome network instability, and affords cleaner restarts of large files. This work will also improve transport speeds by eventually serving as the backbone for multi-target transfers into iRODS and multi-source transfers out of iRODS.</p>
+                        <p>The outcome of the C++ REST API will play a huge part in answering that question.</p>
+                    </div>
 
-                <p>Because each part of a larger file will now be addressable, iRODS will be able to provide POSIX-like semantics for object storage technologies.  This means that these object stores can be connected to iRODS in a cache-free configuration.</p>
+                    <h3 id="new-custom-protocol">New Custom Protocol</h3>
```
Comment (trel) [inline content/pages/roadmap.html:61]: not sure we need an anchor here.

**Review by trel — COMMENTED** [review]
definitely think we should have this page be a historical record of the roadmaps over time...

- Kory 2022
- Terrell 2017
- Jason 2015

Perhaps the 'not current' ones are in a different section or accordion or arrow flip...

We can find the earlier text in the history of this file...

File: content/pages/roadmap.html [hunk 2]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                <p>Making changes to iRODS is difficult. With the release of the 4.3 series, the Consortium strived to improve the implementation by cleaning up semantics. This was accomplished through various techniques. For example, by introducing higher level abstractions, the implementation became easier to reason about while also allowing the internals to be changed without fear of breaking existing deployments. iRODS 4.4 will continue this effort by attempting to address the issues outlined below.</p>
 
-                <h3>Logging</h3>
+                <h2>The iRODS Protocol</h2>
 
-                <p>The logging in iRODS needs to be modernized and standardized with regards to best practices around log rotation, timestamps, and human and machine readable formats.  We plan to investigate a variety of logging frameworks and select one that delivers flexibility powerful enough to afford easy aggregation of logs from our distributed system.</p>
+                <p>iRODS implements its own protocol. It is not documented or versioned. Attempting to add support for other programming languages is a real challenge to developers unfamiliar with iRODS.</p>
```
Comment (trel) [inline content/pages/roadmap.html:45]: "documented or versioned" -> "documented or versioned outside of the implementation itself".

File: content/pages/roadmap.html [hunk 3]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                <div>
+                    <p>To make iRODS more accessible, the Consortium will be investigating the following approaches:</p>
 
-                <p>The everpresent backlog of bugfixes and edge cases will continue to be addressed.  This release will mark the first of a new tick-tock release pattern, meaning one release will focus on features, and the next will focus more on hardening.  4.2 was the feature release, and 4.3 will be the hardening release.</p>
+                    <h3>Single Standardized Protocol</h3>
+                    <div>
+                        <p>The idea of moving to a standardized protocol was motivated by the efforts behind the C++ REST API.</p>
 
-                <h3>Multipart Data Objects</h3>
+                        <p>The HTTP protocol is the primary candidate here. It is simple to understand and adding support for it is literally a matter of choosing a good library. S3 has seen major success and it is built entirely on the HTTP protocol. So, it makes a lot of sense to choose HTTP for iRODS. Doing that instantly widens the reach of iRODS.</p>
```
Comment (trel) [inline content/pages/roadmap.html:54]: "it makes a lot of sense" -> "it may make a lot of sense"

File: content/pages/roadmap.html [hunk 4]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                        <p>The outcome of the C++ REST API will play a huge part in answering that question.</p>
+                    </div>
 
-                <p>Because each part of a larger file will now be addressable, iRODS will be able to provide POSIX-like semantics for object storage technologies.  This means that these object stores can be connected to iRODS in a cache-free configuration.</p>
+                    <h3 id="new-custom-protocol">New Custom Protocol</h3>
+                    <div>
+                        <p>
+                        This approach is about cleaning up the design of the original protocol. From a functionality standpoint, interacting with iRODS would remain nearly identical. The primary difference would be in the design and implementation. The primary benefit of this is that it provides freedom in the sense that it can changed.
```
Comment (trel) [inline content/pages/roadmap.html:64]: "that it can changed" -> "that it can be changed"

File: content/pages/roadmap.html [hunk 5]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                    <h3>New Plugin Interface</h3>
+                    <div>
+                        <p>This approach is more complex, but offers the most flexibility to developers.</p>
 
-                <p>We are working with Intel's Lustre group to investigate and then implement best practices around keeping the iRODS catalog up-to-date for such high-velocity systems.</p>
+                        <p>Introducing a plugin interface for protocols means the number of protocols supported by iRODS is unbounded. However, this requires expert level design skills and a full understanding of all components involved with iRODS and networking.</p>
 
-                <h3>Metadata Abstraction</h3>
+                        <p>This effort must satisfy the same requirements defined under <a href="#new-custom-protocol">New Custom Protocol</a>. Ultimately, this is a final resort. Given that the focus of 4.4 is simplification and maintainability, we prefer to keep the number of protocols at one.</p>
```
Comment (trel) [inline content/pages/roadmap.html:85]: oh, i see the link to the earlier anchor now.

File: content/pages/roadmap.html [hunk 6]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                <h2>Message Encoding</h2>
 
-                <p><a href="https://github.com/xu-hao/QueryArrow/">QueryArrow</a> has been developed to help alleviate this restriction and allow iRODS to store and query metadata in a variety of other database or indexing technologies.</p>
+                <p>iRODS implements its own message encoding via a library call PackStruct. PackStruct supports a binary format and an XML-like format. The only form of documentation for these encodings is the source code. The issue with using the source code as a reference for understanding PackStruct is that PackStruct's implementation is difficult to understand. Given iRODS' history, this is understandable. However, there are several encoding libraries available that provide features such as efficient binary encoding, versioning, and cross-language support. iRODS is at a point where it must switch to using a well maintained message encoding.</p>
```
Comment (trel) [inline content/pages/roadmap.html:91]: "Given iRODS' history" -> "Given iRODS' lengthy history"

File: content/pages/roadmap.html [hunk 7]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
 
-                <p>We plan to release the QueryArrow database plugin which will afford this abstraction of metadata.</p>
+                <p>The Consortium will be introducing a new plugin interface that gives developers a way to experiment with different message encoding libraries. Adding this interface also means the server and clients must learn how to negotiate which encoding to use. The first plugin implementation will be PackStruct.</p>
+
+                <h2>Cloud-Friendly</h2>
+
+                <div>
+                    <p>Cloud environments offer many benefits. One of the primary reasons for running applications in the cloud is scalability. While iRODS can be run in a cloud environment, scaling iRODS horizontally requires the same amount of work as if running iRODS on premise. Getting iRODS to the point where administrators can turn a dail and get more iRODS nodes will require some real thought.</p>
```
Comment (trel) [inline content/pages/roadmap.html:98]: "on premise" -> "on premises"

Comment (trel) [inline content/pages/roadmap.html:98]: "dail" -> "dial"

File: content/pages/roadmap.html [hunk 8]
```diff
@@ -16,61 +16,98 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                    <ul>
+                        <li>Fast startup time</li>
+                        <li>Improved handling of host names</li>
+                        <li>Every server becomes a Provider</li>
+                    </ul>
+                </div>
+
+                <h2>Improved support for Archive files</h2>
+
+                <p>The facilities provided by iRODS to manipulate archive files needs to be modernized. The codebase contains several special code paths for dealing with archive files. This makes the implementation brittle and difficult to reason about. Users and administrators need tools with clear semantics. Therefore, we are planning on reimplementing support for archive files from the ground up.</p>
```
Comment (trel) [inline content/pages/roadmap.html:110]: "manipulate archive files" -> "manipulate archive files in various formats"

**Review by trel — COMMENTED** [review]
i'm going to add a commit to put my name on the 4.2 and resurrect jason's original doc as well.

File: content/pages/roadmap.html (hunk 5 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Do we want to restore the anchor?

File: content/pages/roadmap.html (hunk 2 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html (hunk 3 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html (hunk 4 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html (hunk 6 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html (hunk 7 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html (hunk 8 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: done

File: content/pages/roadmap.html [hunk 9]
```diff
@@ -16,25 +16,146 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
 
                 <p>If you have an opinion about where iRODS should go next, I encourage you to participate in this open software development process. Join the <a href="https://groups.google.com/forum/#!forum/irod-chat">iRODS chat Google Group</a>. Raise issues on <a href="https://github.com/irods/irods">GitHub</a>. And if you are a <a href="{filename}/pages/consortium-membership.html">Consortium Member</a>, get involved in the Technology Working Group.</p>
 
-                <p>Terrell Russell<br/>
+                <p>Kory Draughn<br/>
```
Comment (korydraughn) [inline content/pages/roadmap.html:19]: Perhaps I should restore this?

Seems weird to have my name attached to this section.

File: content/pages/roadmap.html (hunk 5 above)
Comment (trel) [reply content/pages/roadmap.html]: no, the document isn't that heavy/long.   let's keep them text within each section...

Comment (korydraughn) [reply content/pages/roadmap.html]: Ok

File: content/pages/roadmap.html (hunk 9 above)
Comment (trel) [reply content/pages/roadmap.html:19]: no?   you're the chief technologist, and it says 'I' in there... so someone is saying it :)

Comment (korydraughn) [reply content/pages/roadmap.html:19]: Ok

**Comment by trel** [comment]
this is looking pretty good to me...

overview, history sections, toggles... checks the boxes.

**Comment by korydraughn** [comment]
I agree. I still have to go back and tweak things based on review comments that happened outside of github (e.g. framing).

**Comment by trel** [comment]
gotcha

File: content/pages/roadmap.html [hunk 10]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                <h1><i class="green_font fa fa-angle-right"></i> <a onclick="togglevisibility('2023body')">2023</a></h1>
+
+                <div id="2023body" style="display:block">
+
+                    <p>Kory Draughn<br/>
+                    Chief Technologist<br/>
+                    The iRODS Consortium<br/>
+                    <i>January 2023</i></p>
+
+                    <h2>iRODS 20XX</h2>
```
Comment (trel) [inline content/pages/roadmap.html:59]: "20XX" -> "2023 Roadmap"

File: content/pages/roadmap.html [hunk 11]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+
+                    <p>With each release of iRODS, the server, libraries, and clients become more stable and featureful. The needs of users and administrators become more complex over time. Users want tools and GUIs that are easy to understand quickly. Administrators want more flexibility and visibility into what the system is doing.</p>
+
+                    <p>In order to facilitate these needs, the Consortium has three goals: Simplification, Maintainability, and Approachability.</p>
+
+                    <p>We plan to satisfy these goals by doing the following:</p>
```
Comment (trel) [inline content/pages/roadmap.html:65]: Upcoming attention will be paid to the following:

File: content/pages/roadmap.html [hunk 12]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+
+                    <h3>The Server</h3>
+
+                    <ul>
+                        <li>
+                            Make implementing the iRODS Protocol simple by moving to HTTP or designing a new protocol
+                            <ul>
+                                <li>The move to HTTP will be heavily influenced by the C++ REST API</li>
```
Comment (trel) [inline content/pages/roadmap.html:73]: "The move" -> "Any move"

File: content/pages/roadmap.html [hunk 13]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+
+                    <p>In order to facilitate these needs, the Consortium has three goals: Simplification, Maintainability, and Approachability.</p>
+
+                    <p>We plan to satisfy these goals by doing the following:</p>
+
+                    <h3>The Server</h3>
+
+                    <ul>
+                        <li>
+                            Make implementing the iRODS Protocol simple by moving to HTTP or designing a new protocol
```
Comment (trel) [inline content/pages/roadmap.html:71]: "simple" -> "less complicated"

File: content/pages/roadmap.html [hunk 14]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                            </ul>
+                        </li>
+
+                        <li>
+                            Replace the PackStruct serialization library
+                            <ul>
+                                <li>
+                                    Use an actively maintained and proven serialization library (e.g. Protocol Buffers, Flatbuffers, etc.)
+                                    <ul>
+                                        <li>Enables simple path to supporting the most popular programming languages</li>
```
Comment (trel) [inline content/pages/roadmap.html:84]: "Enables simple path" -> "Enable a straightforward"

File: content/pages/roadmap.html [hunk 15]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                                </li>
+                            </ul>
+                        </li>
+
+                        <li>
+                            Make the server cloud-friendly
+                            <ul>
+                                <li>Improve startup time</li>
+                                <li>Improve flexibility around dynamic host names</li>
+                                <li>Every server becomes a Provider</li>
```
Comment (trel) [inline content/pages/roadmap.html:96]: "Every server becomes a Provider" -> "Treat every server as a Catalog Service Provider"

File: content/pages/roadmap.html [hunk 16]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                            </ul>
```
Comment (trel) [inline content/pages/roadmap.html:97]: <li>Centralize configuration</li>

File: content/pages/roadmap.html [hunk 17]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                    </ul>
+
+                    <h3>Clients</h3>
+
+                    <ul>
+                        <li>
+                            Replace iCommands with a new CLI
+                            <ul>
+                                <li>A single binary backed by a plugin architecture allowing its functionality to be extended</li>
+                                <li>No memory leaks and properly handles signals</li>
```
Comment (trel) [inline content/pages/roadmap.html:145]: "properly handles signals" -> "proper signal handling"

File: content/pages/roadmap.html [hunk 18]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                            </ul>
+                        </li>
+
+                        <li>
+                            Improve Metalnx
+                            <ul>
+                                <li>Add support for the <strong>pam_interactive</strong> authentication scheme</li>
+                                <li>Remove the need for helper database</li>
+                                <li>
+                                    Remove all administrative features
```
Comment (trel) [inline content/pages/roadmap.html:156]: "all" -> "most"

**Review by trel — COMMENTED** [review]
suggestions inline

File: content/pages/roadmap.html (hunk 18 above)
Comment (korydraughn) [reply content/pages/roadmap.html:156]: Most? What features would possibly remain?

Comment (trel) [reply content/pages/roadmap.html]: change ownership/permissions on data objects?   queries that know you're an admin... -M type stuff...

File: content/pages/roadmap.html (hunk 10 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 11 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 12 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 13 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 14 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 15 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 16 above)
Comment (korydraughn) [reply content/pages/roadmap.html:98]: Done

File: content/pages/roadmap.html (hunk 17 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html (hunk 18 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

File: content/pages/roadmap.html [hunk 19]
```diff
@@ -16,25 +16,187 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                    <p>Kory Draughn<br/>
+                    Chief Technologist<br/>
+                    The iRODS Consortium<br/>
+                    <i>January 2023</i></p>
+
+                    <h2>iRODS 2023 Roadmap</h2>
+
+                    <p>With each release of iRODS, the server, libraries, and clients become more stable and featureful. The needs of users and administrators become more complex over time. Users want tools and GUIs that are easy to understand quickly. Administrators want more flexibility and visibility into what the system is doing.</p>
+
+                    <p>In order to facilitate these needs, the Consortium has three goals: Simplification, Maintainability, and Approachability.</p>
```
Comment (korydraughn) [inline content/pages/roadmap.html:63]: How about ...

In order to facilitate these needs, the Consortium [is focused / will focus] on improving approachability.

- OR -

In order to facilitate these needs, the Consortium plans to improve approachability by focusing on the following:
_(this replaces L65)_

Thoughts?

**Comment by korydraughn** [comment]
Also, shouldn't we make the 4.1 section expandable/collapsible? All we're missing is a year.

File: content/pages/roadmap.html (hunk 19 above)
Comment (trel) [reply content/pages/roadmap.html:63]: yes, the second is nice.

**Comment by trel** [comment]
No, July 2015 is the first time we had a written roadmap.

File: content/pages/roadmap.html (hunk 19 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done.

**Comment by korydraughn** [comment]
I think this is in a good spot. Any remaining thoughts?

Officially opening up for review.

File: content/pages/roadmap.html [hunk 20]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                    <ul>
+                        <li>Finish and release the Policy Composition rule engine plugin</li>
+                        <li>Finish and release the Publishing capability</li>
+                    </ul>
+
+                    <h3>Libraries</h3>
+
+                    <ul>
+                        <li>
+                            Add support for new Authentication Plugin Framework introduced in the iRODS 4.3 series
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply content/pages/roadmap.html]: Done

**Comment by trel** [comment]
if there's nothing that's changed since voting started - please squash to one commit and we'll hover over the merge button in the coming week.

**Comment by korydraughn** [comment]
What about your commits?

Won't they be lost if I squash everything done?

**Comment by korydraughn** [comment]
Squashed and verified that it matches what was sent out to consortium members.

File: content/pages/roadmap.html [hunk 21]
```diff
@@ -16,25 +16,186 @@ <h4><i class="fa fa-map"></i> <b>Roadmap</b></h2>
...
+                <div id="2023"></div>
+
+                <h1><i class="green_font fa fa-angle-right"></i> <a onclick="togglevisibility('2023body')">2023</a></h1>
+
+                <div id="2023body" style="display:block">
+
+                    <p>Kory Draughn<br/>
+                    Chief Technologist<br/>
+                    The iRODS Consortium<br/>
+                    <i>February 2023</i></p>
```
Comment (trel) [inline content/pages/roadmap.html:57]: Let's change to April.

Comment (korydraughn) [inline content/pages/roadmap.html:57]: Should the month be changed?

**Review by trel — COMMENTED** [review]
i'd like to see a render one more time.

File: content/pages/roadmap.html (hunk 21 above)
Comment (korydraughn) [reply content/pages/roadmap.html]: Done.

**Comment by trel** [comment]
tweak the commit message... perhaps to... 'Added 2023 Roadmap'

**Comment by korydraughn** [comment]
Commit message updated.

**Comment by trel** [comment]
render looks good.   will merge sometime next week.

**Review by trel — APPROVED** [review]

**Comment by korydraughn** [comment]
I updated the README to include a section explaining how to render the site locally.

I've also verified that each line, word for word, matches what was sent to members.

I think we are good to go here.

File: README.md [hunk 22]
```diff
@@ -45,3 +45,11 @@ rsync -arv output/ user@target:/path/to/output/
 ```
 
 - For Apache, allow the deployed `.htaccess` to override settings for the active `Directory`
+
+## View
+
+You can view the output using Docker by running the following:
+
+```
+docker run -dit --rm --name irods.org -p 9090:80 -v $(pwd)/output:/usr/local/apache2/htdocs/ httpd:2.4
```
Comment (trel) [inline README.md:54]: it's either `-d` or `-it`, right?

not both?

Comment (korydraughn) [reply README.md:54]: I just copied that from the irods_docs repo. I think `-d` is all that's needed.

Let me confirm that real quick.

Comment (korydraughn) [reply README.md]: Done.

**Comment by korydraughn** [comment]
Yes.

**Pull request merged**